src/lib.rs

```diff
@@ -384,21 +384,50 @@ impl Drop for Context {
 
 pub struct Socket {
     sock: *mut c_void,
-    closed: bool
+    owned: bool,
 }
 
 unsafe impl Send for Socket {}
 
 impl Drop for Socket {
     fn drop(&mut self) {
-        match self.close_final() {
-            Ok(()) => { debug!("socket dropped") },
-            Err(e) => panic!(e)
+        if self.owned {
+            if unsafe { zmq_sys::zmq_close(self.sock) } == -1i32 {
+                panic!(errno_to_error());
+            } else {
+                debug!("socket dropped");
+            }
         }
     }
 }
 
 impl Socket {
+    /// Consume the Socket and return the raw socket pointer.
+    ///
+    /// Failure to close the raw socket manually or call `from_raw`
+    /// will lead to a memory leak.
+    pub fn to_raw(mut self) -> *mut c_void {
+        self.owned = false;
+        self.sock
+    }
+
+    /// Create a Socket from a raw socket pointer. The Socket assumes
+    /// ownership of the pointer and will close the socket when it is
+    /// dropped.
+    pub fn from_raw(sock: *mut c_void) -> Socket {
```

Comment:
This should be unsafe as in https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw